### DISCUSSION TITLE
Complete SBOM? (IDFGH-11319)

### DISCUSSION THREAD

**Discussion by fpgax** [espressif/esp-idf-sbom#8] [General]
Previously in another issue I received an example SBOM.
I also asked previously about mbed-tls in yet another issue.

Now, I've been analyzing it a bit and I have a few additional questions.

The mbed-tls is expressed in the example SBOM as two separate things:

mbedtls component (Espressif)
and
mbedtls-mbedtls submodule (ARM)

Now, I think that in general the Espressif-component actually shouldn't be in the SBOM, since it has no CPE. Nobody will be cross-checking this component towards the NVD for example. It should get folded up into esp-idf which has a CPE. Registering a specific CPE for this subcomponent, which is only used by esp-idf would also not be an appropriate solution.

Any thoughts on this?

The second part of my question relates to another component, FreeRTOS. I can't from the example-SBOM figure out the base version of FreeRTOS.

So this feeds into my second request, regarding f.ex. FreeRTOS, but really the SBOM in general: Are you planning to generate a complete SBOM for any SDK version? And in such case any idea when?

In a complete SBOM I would expect to see exact versions of components, like for example for mbed-tls and FreeRTOS, but not have "unnecessary" components in there either, like for tooling, testing, internal components, and so forth. Just information which has relevance for cybersecurity and in a format which is typically useful for the cybersecurity tasks, e.g. cross-checking against a database.

Have you thought about this more general problem statement? Are you in a position to comment on your plans going forward?

**Comment by andyn-ff** [comment]
Sharing a complete SBOM in for new releases of ESP-IDF would be a _great_ step while you are still adding `sbom.yaml` files to the ESP-IDF components themselves to enable automatic generation of such a SBOM by the esp-idf-sbom tool. If we had a static SBOM in SPDX format to represent the released version of ESP-IDF, we could add sbom.yaml files to our own project components and use esp-idf-sbom to generate a partial SBOM for our 'user' components, then merge with the static SBOM from Espressif.

**Comment by fhrbata** [comment]
Hello @fpgax,

>The mbed-tls is expressed in the example SBOM as two separate things:
>
>mbedtls component (Espressif)
>and
>mbedtls-mbedtls submodule (ARM)
>
>Now, I think that in general the Espressif-component actually shouldn't be in the SBOM, since it has no CPE. Nobody >will be cross-checking this component towards the NVD for example. It should get folded up into esp-idf which has a >CPE. Registering a specific CPE for this subcomponent, which is only used by esp-idf would also not be an appropriate >solution.
>
>Any thoughts on this?

This is by design. The whole build process is based on components and their dependencies. This is mirrored into the generated sbom, where each component is represented with `SPDX package`. In general, component may or may not include 3rd party libs. The mbedtls component just wraps the upstream mbedtls library, so for this particular case it would make sense to skip the component and refer to the library alone. But we need some more generic approach, which can handle all components automatically, hence this layout. I could be missing something, but I personally don't see problem with this approach.

IMHO it's not all about which component/lib/tool has a CPE. It's important to know what was used to build the final binary. Let's say that I generate sbom for some project, which uses component without CPE and later a vulnerability is found and CVE/CPE created. This could be even for Espressif's specific component not using any 3rd party lib. I would like to go back to check my sbom if the vulnerable component was used, even though by the time the sbom was generated there was not CPE. Also note that while `esp-idf-sbom` currently checks against NVD based on CPEs only, there are other tools like [cve-bin-tool](https://github.com/intel/cve-bin-tool), which can check the sbom against NVD and  other databases based on name/version. 


>The second part of my question relates to another component, FreeRTOS. I can't from the example-SBOM figure out >the base version of FreeRTOS.

I guess your are referring to the example from #2 . It was generated when there was no manifest file for freertos. It should be added now. For example sbom generated from `esp-idf` master branch for `hello_world` has it.

```
# freertos component
PackageName: component-freertos
SPDXID: SPDXRef-COMPONENT-freertos
PackageVersion: v5.2-dev-3829-gc5ae121bacec
PackageSupplier: Organization: Espressif Systems (Shanghai) CO LTD
PackageDownloadLocation: NOASSERTION
FilesAnalyzed: false
PackageLicenseConcluded: NOASSERTION
PackageLicenseDeclared: NOASSERTION
PackageCopyrightText: NOASSERTION
Relationship: SPDXRef-COMPONENT-freertos DEPENDS_ON SPDXRef-SUBPACKAGE-freertos-FreeRTOS-Kernel

# freertos-FreeRTOS-Kernel subpackage
PackageName: freertos
PackageSummary: <text>An open-source, real-time operating system (RTOS) with additional features and patches from Espressif.</text>
SPDXID: SPDXRef-SUBPACKAGE-freertos-FreeRTOS-Kernel
PackageVersion: 10.5.1
PackageSupplier: Organization: Espressif Systems (Shanghai) CO LTD
PackageOriginator: Organization: Amazon Web Services
PackageDownloadLocation: NOASSERTION
FilesAnalyzed: false
PackageLicenseConcluded: NOASSERTION
PackageLicenseDeclared: NOASSERTION
PackageCopyrightText: NOASSERTION
ExternalRef: SECURITY cpe23Type cpe:2.3:o:amazon:freertos:10.5.1:*:*:*:*:*:*:*
```

>So this feeds into my second request, regarding f.ex. FreeRTOS, but really the SBOM in general: Are you planning to >generate a complete SBOM for any SDK version? And in such case any idea when?

I have no information about such plan and I'm not sure it's actually need, but I can be missing something. The idea is that you checkout/use any esp-idf version for your project and the sbom will be generated based on the actual build information and components used/linked into the final binary. The point is that usually the final binary does not include everything from the SDK, but just a set of components linked into it. I think it's also important to avoid false positive reports. Let's say that this complete SDK SBOM contains a component with a vulnerability, but your project is not actually using/linking that component. I think in such cases users don't want their project to be flagged with that vulnerability. 

>In a complete SBOM I would expect to see exact versions of components, like for example for mbed-tls and FreeRTOS, >but not have "unnecessary" components in there either, like for tooling, testing, internal components, and so forth. >Just information which has relevance for cybersecurity and in a format which is typically useful for the cybersecurity >tasks, e.g. cross-checking against a database.

We are constantly improving and adding the metadata with manifest files, which are used to generate the sbom. So more and more components will have the correct versions. I think even now we have most of the mostly exposed components covered.

EDIT: I forged to add that you can trim down the components in the sbom with `--rem-unused` option. It will remove all components, which are not linked into the final binary.

>Have you thought about this more general problem statement? Are you in a position to comment on your plans going >forward?

I think the overall goal at this point is to have manifest files add for critical components, so the sbom contains valid and accurate information. As I said previously we are constantly adding manifest. I hope this can be already seen in our repositories.

Thank you!

**Comment by fhrbata** [comment]
> Sharing a complete SBOM in for new releases of ESP-IDF would be a _great_ step while you are still adding `sbom.yaml` files to the ESP-IDF components themselves to enable automatic generation of such a SBOM by the esp-idf-sbom tool. If we had a static SBOM in SPDX format to represent the released version of ESP-IDF, we could add sbom.yaml files to our own project components and use esp-idf-sbom to generate a partial SBOM for our 'user' components, then merge with the static SBOM from Espressif.

As mentioned in my previous reply. I think this would lead to false positive reports. The static SBOM would contain all components even though your project is not using them. Meaning your firmware may be flagged as vulnerable even though it doesn't use the vulnerable component.

Again thank you for your feedback it's much appreciated!

**Comment by fhrbata** [comment]
Hello @fpgax , does the above comment answer your question(address your concern)? Thank you

**Comment by fpgax** [comment]
First of all: Thanks! You're doing great work here. We are using the esp-idf SDK in multiple products and your continued work on SBOM is of great help.

You are completely right about not including unused components. It was no my intention to hint at anything else. When I say complete and correct, I mean exactly to neither miss anything nor to include something which isn't there. So that part we totally agree on!

For the products that we build using the SDK, we also naturally will need to create an SBOM. The main policy driver for what should be included in an SBOM is the question: Is it cybersecurity relevant? (Licensing, the other major usage for SBOMs, is something I don't concern myself with)

There is of course an internal structure to esp-idf, but to the largest extent the only relevant usage is to cross-check it with a database, like the NVD. For that purpose it is irrelevant what the internal code structure or architecture of esp-idf is. For example that there is a git submodule, which itself has 3 components and they in turn include some libraries. This is all mostly irrelevant information when running a vulnerability scanner. The SBOM isn't the place to start to investigate the structure of the code base and to look for architectural information (in my opinion that is). That's what the source code repo is for.

If these open source libraries however have or should have a CPE, meaning some 3rd party library or even an internal library clearly distinguishable from esp-idf, for example shared between different internal projects, then indeed it becomes relevant. Otherwise, for me, all that structure just folds right up into esp-idf itself and it's sufficient in the SBOM to state the esp-idf version and everything else follows from that. Any CVEs in internal subcomponents/libraries, unless they're clearly separated and with a CPE of their own, will be reported on esp-idf.

So that's why for example esp-idf, mbed-tls and freertos are cybersecurity relevant for the SBOM, but for example cxx or pthread (if I understand correctly that these components are purely esp-idf internal) aren't.

I don't think it's wrong as such to include more information and structure, but for our products we will likely take the minimalistic approach and just focus on cybersecurity relevance (i.e. to enable vulnerability scanning).

Having this discussion with someone actively working on the same type of problem statement is indeed very useful for me, even if we end up taking different approaches. I would still love to hear your thoughts!

**Comment by fhrbata** [comment]
Hello @fpgax ,

> First of all: Thanks! You're doing great work here. We are using the esp-idf SDK in multiple products and your continued work on SBOM is of great help.

Thank you and many thanks for your feedback and input. It's very helpful!

> You are completely right about not including unused components. It was no my intention to hint at anything else. When I say complete and correct, I mean exactly to neither miss anything nor to include something which isn't there. So that part we totally agree on!

Ah, it's possible I misread your comment. My apologies.

> For the products that we build using the SDK, we also naturally will need to create an SBOM. The main policy driver for what should be included in an SBOM is the question: Is it cybersecurity relevant? (Licensing, the other major usage for SBOMs, is something I don't concern myself with)

I understand your primary interest is security and vulnerability scanning of  SBOM. As you stated it's one of SBOM usage. But there might be other use cases, which might be of interest to others, like supply chain.


> There is of course an internal structure to esp-idf, but to the largest extent the only relevant usage is to cross-check it with a database, like the NVD. 

IIUC, component relationships and supply chain is one of the key aspects of SBOM. There are several  
[NTIA SBOM](https://www.ntia.gov/page/software-bill-materials) documents that explicitly mention this. For example in [SBOM at a Glance](https://www.ntia.gov/sites/default/files/publications/sbom_at_a_glance_apr2021_0.pdf)

_What is an SBOM?
An SBOM is a formal, machine-readable inventory of software components and dependencies,
information about those components, and their hierarchical relationships. ..._


>For that purpose it is irrelevant what the internal code structure or architecture of esp-idf is. For example that there is a git submodule, which itself has 3 components and they in turn include some libraries. This is all mostly irrelevant information when running a vulnerability scanner. The SBOM isn't the place to start to investigate the structure of the code base and to look for architectural information (in my opinion that is). That's what the source code repo is for.

I guess it depends on what you are looking for. Let's say that there is a vulnerability in one of the 3rd party libraries. While evaluating the vulnerability and its impact, you might be interested to know how it got into your application and through which esp-idf component/git submodule/library chain it was. You can analyze the sources and get to the same information manually, but SBOM can help with this.

For example this [sbom.pdf](https://github.com/espressif/esp-idf-sbom/files/13251176/sbom.pdf) was generated with following commands.

```
# generate minimal SBOM for hello_world example and remove all components not directly involved in the final binary
$ esp-idf-sbom create --rem-config --rem-unused -o sbom.spdx build/project_description.json
# generates dot file  showing relationships within SBOM
$ sbom2dot -i sbom.spdx -o sbom.dot
# generate pdf from dot
$ dot -Tpdf -o sbom.pdf sbom.dot
```   

> If these open source libraries however have or should have a CPE, meaning some 3rd party library or even an internal library clearly distinguishable from esp-idf, for example shared between different internal projects, then indeed it becomes relevant. Otherwise, for me, all that structure just folds right up into esp-idf itself and it's sufficient in the SBOM to state the esp-idf version and everything else follows from that. Any CVEs in internal subcomponents/libraries, unless they're clearly separated and with a CPE of their own, will be reported on esp-idf.

My understanding is that typically CPE is not created until a CVE is filed. Meaning if there is no CVE for some software product, there will be no CPE.

Let's say I generate SBOM for my application, which will link some upstream `somelib.a` and at the time my application is built there is no CPE record for `somelib.a`. I ship the application to customers and later someone finds a vulnerability in `somelib.a` version `1.2.3` and CVE/CPE is created. If there is no SPDX package for `somelib.a`, customers will not know about the vulnerability. This can IMHO also happen to e.g. esp-idf component, which modifies some upstream library, when a vulnerability is found just in the modified code and not in the upstream version. Maybe the vulnerability could be tracked under the already existing `espressif:esp-idf` CPE or a new CPE, just for that esp-idf component, may be created.

If the package and it's version is present in SBOM, even if it has no CPE, it still can be scanned. Note that currently `esp-idf-sbom` checker works only with CPE information, but there are other tools like [cve-bin-tool](https://github.com/intel/cve-bin-tool), which do vulnerability scanning based on package name and version. Meaning you can still check the package, even if it doesn't have any CPE at the build time. We should probably add this type of scanning to `esp-idf-sbom` tool too in the future.

> So that's why for example esp-idf, mbed-tls and freertos are cybersecurity relevant for the SBOM, but for example cxx or pthread (if I understand correctly that these components are purely esp-idf internal) aren't.
>
> I don't think it's wrong as such to include more information and structure, but for our products we will likely take the minimalistic approach and just focus on cybersecurity relevance (i.e. to enable vulnerability scanning).

Since there might be different use cases for the generated SBOM, we tried to make the information, which will be included in SBOM, configurable. Currently the absolute minimal SBOM `esp-idf-sbom` can generate is with the following command

```
$ esp-idf-sbom create --rem-config --rem-unused -o sbom.spdx build/project_description.json
```
But it will contains SPDX packages for esp-idf components through which the 3rd party libraries were included and of course components, which are included/linked into the final binary.

Do you think this minimal SBOM can help in your use case? I guess we can add an option to include only components/libs, which have CPE already, but as I stated above, I think we can never know when/how a CPE will be created.

> 
> Having this discussion with someone actively working on the same type of problem statement is indeed very useful for me, even if we end up taking different approaches. I would still love to hear your thoughts!

Likewise, it's always great to hear thoughts and ideas from others and thank for for that!

**Comment by fpgax** [comment]
> Let's say I generate SBOM for my application, which will link some upstream somelib.a and at the time my application is built there is no CPE record for somelib.a. I ship the application to customers and later someone finds a vulnerability in somelib.a version 1.2.3 and CVE/CPE is created.

But how can a scanner understand what somelib.a is without a CPE? Wouldn't it be better to try to register a CPE even without any CVE? If NIST doesn't allow that currently, then I think they could be convinced to change their stance.

**Comment by fhrbata** [comment]
Hello @fpgax ,

I think the point here is that some scanners are not using just CPEs. They query the NVD based on package name and version, which is part of the SPDX SBOM. So for the `somelib.a` example there will be SPDX package in the sbom. Something like this

```
PackageName: somelib
PackageVersion: 1.2.3
```

This should be enough information e.g. for [cve-bin-tool](https://github.com/intel/cve-bin-tool) to do the scanning. The scanner can use this information to construct the CPE(`cpe:2.3:a:*:somelib:1.2.3:*:*:*:*:*:*:*`) or can have some internal database where it can look based on the name/version.  If there is no CPE, there is no vulnerability. But once the CPE is filed, it should be reported. 

Thank you

**Comment by fpgax** [comment]
Right, I see your point, although it will be a little bit of a guess regarding the specific identifier? In some sense we might as well just guess the CPE itself. I think a conversation needs to be had with NIST regarding this problem, so that CPEs can be registered beforehand, in case they don't allow it today.

**Comment by fhrbata** [comment]
Hello @fpgax ,

> Right, I see your point, although it will be a little bit of a guess regarding the specific identifier? 

Yes, I guess this can lead to some ambiguity.

>In some sense we might as well just guess the CPE itself. I think a conversation needs to be had with NIST regarding this problem, so that CPEs can be registered beforehand, in case they don't allow it today.

Here is some documentation about the CPE registration process.

[https://cpe.mitre.org/dictionary/](https://cpe.mitre.org/dictionary/)
_CPE Names (also called CPE Identifiers, CPE-IDs, and CPEs) are created by the CPE Community on an as-needed basis, and must be well-formed as determined by the appropriate CPE Specifications. Community members who require a new CPE Name should create a CPE-conformant name and submit it to the CPE Team for inclusion in the CPE Dictionary. The process is as follows:_

[https://nvd.nist.gov/products/cpe](https://nvd.nist.gov/products/cpe)
_As of December 2009, The National Vulnerability Database is now accepting contributions to the Official CPE Dictionary. Organizations interested in submitting CPE Names should contact the NVD CPE team at cpe_dictionary@nist.gov for help with the processing of their submission._

IIUC how the [CPE matching](https://nvlpubs.nist.gov/nistpubs/Legacy/IR/nistir7696.pdf) works, a CPE is created for specific version of software for which a vulnerability is found. IOW there is no need to have CPE for each version. OTOH I expect that once there is some CPE for a product registered, they will use the same ` <part>:<vendor>:<product>:` part. Meaning no need for guessing this part of CPE.

Anyway I'm not sure if it's possible to ask for a CPE registration, for e.g. some 3rd party library, if you are not the author, specially if there is CVE.

I don't think there is nothing much we can do about this.

I managed to allow [discussions](https://github.com/espressif/esp-idf-sbom/discussions) for this project. Would you mind if I convert this issue into a discussion? We can continue this there if necessary?

Thank you

**Comment by fhrbata** [comment]
I'm converting this to a discussion and we can continue the discussion there if necessary. Thank you.

**Comment by fpgax** [comment]
Yes, it's perfect to leave this as a discussion, thanks. I have nothing further at this moment, but perhaps someone else will chime in over time or we'll pick up the continued discussion as this matures in the industry. I see everyone scrambling on this topic at the very least in the EU and the US now, so it's quite good to have these open discussions and learn together.
